ztp/siteconfig-generator/siteConfig/siteConfigBuilder_test.go

```diff
@@ -291,8 +294,105 @@ func Test_getExtraManifestTemplatedRoles(t *testing.T) {
+`
+	tests := []struct {
+		what           string
+		input          string
+		nodeId         int
+		expectedResult string
+		expectedError  string
+	}{{
+		what:           "No template (string)",
+		input:          `key: value`,
+		expectedResult: `key: value`,
+	}, {
+		what:           "No template (int)",
+		input:          `key: 42`,
+		expectedResult: `key: 42`,
+	}, {
+		what:           "Simple template: Site",
+		input:          `key: "{{ .Site.BaseDomain }}"`,
+		expectedResult: `key: "example.com"`,
+	}, {
+		what:           "Simple template: Cluster",
+		input:          `key: "{{ .Cluster.ClusterName }}"`,
+		expectedResult: `key: "cluster1"`,
+	}, {
+		what:           "Simple template: Node",
+		input:          `key: "{{ .Node.HostName }}"`,
+		expectedResult: `key: "node1"`,
+	}, {
+		what:           "Empty string value",
+		input:          `key: "{{ .Site.SshPublicKey }}"`,
+		expectedResult: ``,
+	}, {
+		what:           "Empty slice value",
+		input:          `key: "{{ .Cluster.MachineNetwork }}"`,
+		expectedResult: ``,
+	}, {
+		what:           "Empty struct value",
+		input:          `key: "{{ .Site.PullSecretRef }}"`,
+		expectedResult: ``,
+	}, {
+		// TODO: This should be an error, not a success that returns a nil map
+		what:           "Invalid Node template (no node ID provided)",
+		input:          `key: "{{ .Node.HostName }}"`,
+		nodeId:         -1,
+		expectedResult: ``,
+	}, {
+		what:          "Unparsable key",
+		input:         `key: "{{ Good luck! }}"`,
+		nodeId:        -1,
+		expectedError: "could not be translated",
+	}, {
+		// TODO: This should be an error, not a success that returns a nil map
+		what:           "Invalid key (missing field)",
+		input:          `key: "{{ .Site.NoSuchKey }}"`,
+		expectedResult: ``,
+	}, {
+		// TODO: This should be an error, not a success that returns a nil map
+		what:           "Invalid key (going too deep on a leaf)",
+		input:          `key: "{{ .Node.HostName.Too.Deep }}"`,
+		expectedResult: ``,
+	}, {
+		what:           "Nested structure (no templates)",
+		input:          `top: { middle: { a: "value", b: 42} }`,
+		expectedResult: `top: { middle: { a: "value", b: 42} }`,
+	}, {
+		what:           "Nested structure with templates",
+		input:          `top: { a: "{{ .Site.BaseDomain }}", middle: { b: "{{ .Cluster.ClusterName }}", bottom: "end" } }`,
+		expectedResult: `top: { a: "example.com", middle: { b: "cluster1", bottom: "end" } }`,
+	}, {
+		what:           "Nested structure appending another nested structure",
+		input:          `top: { a: "{{ .Cluster.ClusterLabels }}", middle: { bottom: "end" } }`,
+		expectedResult: `top: { a: {common: "true", "group-du-sno": "", "sites": "test-site"}, middle: { bottom: "end" } }`,
+	}, {
+		// TODO: This should be an error, not a success where the key is removed
+		what:           "Nested recursive structure with invalid key",
+		input:          `top: { middle: { bottom: "{{ .Cluster.NoSuchKey }}" } }`,
```

Comment:
Is the substitution relatively immune to whitespace as well? Might be good to add a few test cases like
`
key: "  {{ .Cluster.Name}}"
key: "{{ .Cluster.Name}}    "
key: "     {{ .Cluster.Name}}    "
`
etc